lumicks/pylake/kymotracker/detail/msd_estimation.py

```diff
@@ -447,3 +455,259 @@ def determine_optimal_points(frame_idx, coordinate, max_iterations=100):
+) -> tuple:
+    """Covariance based estimator.
+
+    The covariance-based diffusion estimator provides a simple unbiased estimator of diffusion.
+    This estimator was introduced in the work of Vestergaard et al [1]. The correction for
+    missing data was introduced in [2].
+
+    Note that this estimator is only valid in the case of pure diffusion without drift.
+
+    References
+    ----------
+    .. [1] Vestergaard, C. L., Blainey, P. C., & Flyvbjerg, H. (2014). Optimal estimation of
+           diffusion coefficients from single-particle trajectories. Physical Review E, 89(2),
+           022726.
+    .. [2] Vestergaard, C. L. (2016). Optimizing experimental parameters for tracking of diffusing
+           particles. Physical Review E, 94(2), 022401.
+
+    Parameters
+    ----------
+    frame_indices : array_like
+        Frame indices
+    x : array_like
+        Positions over time
+    dt : float
+        Time step
+    blur_constant : float
+        Motion blur coefficient. Should be between 0 and 1/4.
+
+        The normalized shutter function is defined as c(t), where c(t) represents whether the
+        shutter is open or closed. c(t) is normalized w.r.t. area. For no motion blur,
+        c(t) = delta(t_exposure), whereas for a constantly open shutter it is defined as
+        c(t) = 1 / dt.
+
+        With c(t) defined as before, the motion blur constant is defined as:
+
+        .. math::
+            R = \frac{1}{\Delta t} \int_{0}^{\Delta t}S(t) \left(1 - S(t)\right)dt
+
+        with
+
+        .. math::
+            S(t) = \int_{0}^{t} c(t') dt'
+
+        When there is no motion blur, we obtain: R = 0, whereas a continuously open shutter over the
+        exposure time results in R = 1/6. Note that when estimating both localization uncertainty
+        and the diffusion constant, the motion blur factor has no effect on the estimate of the
+        diffusion constant itself, but it does affect the calculated uncertainties. In the case of
+        a provided localization uncertainty, it does impact the estimate of the diffusion constant.
+    variance_loc : float
+        Variance of the localization error if it has been determined independently.
+    variance_variance_loc : float
+        The variance of the localization variance estimate (needed if variance_loc is provided).
+
+    Note
+    ----
+    With the SNR defined as:
+
+    .. math::
+        \sqrt{D \Delta t} / \sigma
+
+    With D the diffusion constant and :math:`\sigma`
+    This method requires SNR > 1. Below SNR = 1 it degrades rapidly.
+    """
+    if not 0 <= blur_constant < 0.25:
+        raise ValueError("Motion blur constant should be between 0 and 1/4")
+
+    # Determine the average step size
+    average_frame_step = np.mean(np.diff(frame_indices))
+    avg_dt = average_frame_step * dt
+
+    dx = np.diff(x)
+    mean_dx_squared = np.mean(dx**2)
+
+    if len(x) < 3:
+        raise RuntimeError("Insufficient intervals for using the CVE")
```

Comment:
maybe move towards the top and just throw right away?

Reply:
👍